---
title: "*Xenopus tropicalis* juveniles liver DNA Methylation"
subtitle: "Analysis of differential methylation in *X. tropicalis* frogs liver after exposure to the fungicide Imazalil"
author: 
  - name: Mauricio Roza
    email: "Mauricio.Roza@aces.su.se"
    orcid: 0000-0003-2795-9193
    github: mauricioroza
    affiliations:
    - name: Stockholm University
      department: Department of Environmental Science
      city: Stockholm
      country: Sweden
      url: https://www.su.se/department-of-environmental-science/
    - name: Science for Life Laboratory
      city: Solna
      country: Sweden
      url: https://www.scilifelab.se/

date-modified: last-modified
format: 
  html:
    grid:
      sidebar-width: 300px
      body-width: 900px
      margin-width: 300px
    code-fold: true
    code-tools: true
    highlight-style: github
    toc: true
    toc-depth: 4
    number-sections: true
    number-depth: 4
theme:
  light: flatly
  dark: darkly

execute:
  cache: true
  
title-block-banner: "#002F5F"
css: styles.css

params:
  min_n: 4
  meth_cut: 10
  qvalue_cut: 0.05
  group: "male"
  analysis: "CpGs"
  tiling_window: ""
  tiling_step: ""
---

::: column-body
```{=html}

<div>
    <a href="https://www.linkedin.com/in/mauricio-roza-28b09313b/" target="_blank"><img src="images/linkedin.png" alt="LinkedIn" width="32" height="32"></a>
    <a href="https://github.com/mauricioroza" target="_blank"><img src="images/github.png" alt="GitHub" width="32" height="32"></a>
    <a href="https://scholar.google.com/citations?user=KlosFPIAAAAJ&hl" target="_blank"><img src="images/google_scholar_icon.png" alt="Google Scholar" width="32" height="32"></a>
</div>
```
:::

\ 
 \ 
 
| Analysis Parameters  |                          |
|----------------------|--------------------------|
| Group                | `r params$group`         |
| Min N per group      | `r params$min_n`         |
| Methylation Cut      | `r params$meth_cut`%     |
| q value Cut          | `r params$qvalue_cut`    |
| CpG, tiles or region | `r params$analysis`      |
| Tile Window          | `r params$tiling_window` |
| Tile Step            | `r params$tiling_step`   |

# Analysis Pipeline

```{mermaid}
%%| output-location: column
flowchart TB
  A(Read Coverage Files) --> B(Quality Control)
  B -.-> C(Filtering)
  C --> E((Differential Methylation))
  E --> F(Genome Annotations)
  F --> G(Over-representation Analysis)
  G --> H(Gene Ontology)
  G --> I(KEGG)
  
  subgraph plots[Plots]
    direction TB
    D(Data Structure)
    Q(Volcano)
    R(Circular)
    go(GO)
    K(KEGG)
  end
  
  C --> D
  E --> Q
  E --> R
  H --> go
  I --> K
  
```

::: {.callout-note collapse="true"}
# Read Bismark Coverage Files

Here the Bismark Coverage files paths and sample names are organized in groups for further analysis. `read_bismark_coverage` function wraps up the reading and make it easier to change the possibilities.

```{r}
#| echo: true
#| eval: true 
#| file: code/read_bismark_coverage.R
#| warning: false
#| collapse: false
```

**Select the groups for analysis**

Here I set the `analysis_name` to show up in the graphs titles further in the pipeline. `myobj` sets which groups are being compared. In this case, male control vs male IMZ

```{r}
#| warning: false
file_path_name <- paste0(params$group, "_cut_", params$meth_cut, "_", params$analysis, params$tiling_window)


if (params$group == "male") {
  myobj <- read_bismark_coverage(control_male_path, control_male_sample, imz_male_path, imz_male_sample)
  name_obj <- "Male"
}

if (params$group == "female") {
  myobj <- read_bismark_coverage(control_female_path, control_female_sample, imz_female_path, imz_female_sample)
  name_obj <- "Female"
}

if (params$group == "all") {
  myobj <- read_bismark_coverage(control_path, control_sample, imz_path, imz_sample)
  name_obj <- "All Ctr vs IMZ"
}

if (params$group == "male_vs_female") {
  read_bismark_coverage(male_path, male_sample, female_path, female_sample)
  name_obj <- "All Male vs Female"
}
```

```{r}
if (params$analysis == "promoter") {
  source("code/promoters_analysis.R")
  analysis_object <- "Promoters"
  analysis_name <- paste0("Diff. Methylated Promoters: ", name_obj)

  } else if (params$analysis == "tiles") {
  myobj <- tileMethylCounts(myobj, win.size = params$tiling_window, step.size = params$tiling_step)
  analysis_object <- "DMRs"
  analysis_name <- paste0("Diff. Methylated Regions: ", name_obj)

  } else {
  analysis_object <- "CpG sites"
  analysis_name <- paste0("Diff. Methylated CpGs: ", name_obj)
}
```
:::

::: {.callout-note collapse="true"}
# Quality Control

Histograms showing the % CpG methylation and read coverage.

Ideally the methylation histogram will show two peaks at 0 and 100%, and the coverage read shows an approximate normal distribution.\
A peak of frequency of bases with high coverage indicate PCR bias.

```{r}
#| echo: true
#| eval: true 
#| file: code/qc.R
#| warning: false
#| collapse: true
#| fig-width: 10
#| fig-height: 10
#| message: false
#| results: hide
```
:::

::: {.callout-note collapse="true"}
# Filtering

```{r}
min_n <- params$min_n
```

```{r}
#| echo: true
#| eval: true 
#| file: code/filtering.R
#| warning: false
#| collapse: false
```

Discards bases with coverage below 10 reads and above 99.9th percentile.\
Keep `r analysis_object` with at least **`r min_n`** observations per group.\
After uniting the `r analysis_object`, **`r n_cpg`** are kept.

A further filtering is applied, keeping only `r analysis_object` with standard deviations larger than **2%**, as those are not informative and only increase multiple comparison penalties.\
**`r n_cpg_filtered`** `r analysis_object` are used for further analysis.
:::

# Data Structure

```{r}
#| echo: true
#| eval: true 
#| file: code/data_structure.R
#| warning: false
#| collapse: false
#| output: false
```

-   <font size="6"> **`r n_cpg_filtered`** `r analysis_object` analysed </font>

## Correlation Heatmap

```{r}
#| echo: false
#| eval: true
#| fig-width: 10
#| fig-height: 10

corrplot
```

## Hierarchical Clustering

```{r}
#| echo: false
#| eval: true
#| fig-width: 10
#| fig-height: 10

cluster_meth()
```

## PCA

```{r}
#| echo: false
#| eval: true
#| fig-width: 10
#| fig-height: 10

PCA_meth()
```

# Differential Methylation Analysis

## Code

```{r}
meth_cut <- params$meth_cut
qvalue_cut <- params$qvalue_cut
```

```{r}
#| echo: true
#| eval: true 
#| file: code/differential_methylation.R
#| warning: false
#| collapse: false
#| output: false
```

::: {.callout-note collapse="true"}
# p-value Histogram

```{r}
#| echo: true
#| eval: true
#| fig-width: 10
#| fig-height: 10

par(mfrow = c(2, 2))
hist(meth_diff$pvalue, main = "p-value Distribution")
hist(meth_diff$qvalue, main = "q-value Distribution")
hist(meth_diff$meth.diff, main = "% Methylation Distribution")
```
:::

```{r}
#| echo: true
#| eval: true 
#| file: code/genomic_location_DMR.R
#| warning: false
#| collapse: false
#| output: false
```

```{r}
#| echo: true
#| eval: true 
#| file: code/gene_annotation_overlap_DMR.R
#| warning: false
#| collapse: false
#| output: false
```

## Results

`r analysis_object` were considered differentially methylated when the FDR corrected p-value (q-value) was lower than **`r  qvalue_cut`** and the methylation difference compared with control was higher than **`r  meth_cut`%**

From **`r nrow_meth_diff`** `r analysis_object` analysed, **`r nrow_meth_diff_cut`** are differentially methylated and overlapped with **`r feature_sum_annot$feature |> unique() |> length()`** genes.

**`r gen_regions$promoter`** are located in promoters, **`r gen_regions$exon`** in exons, **`r gen_regions$intron`** in introns, and **`r gen_regions$intergenic`** in intergenic regions.

::: {.callout-note collapse="true"}
# Differentially methylated `r analysis_object` table

```{r}
#| warning: false
library(writexl)

feature_sum_annot %>% 
  dplyr::select(feature, external_gene_name, description, mcols.meth.diff, mcols.pvalue, mcols.qvalue, 
                seqnames, feature_strand, location, start, end) %>%
  datatable() %>%
  formatRound(columns=c('mcols.meth.diff', 'mcols.pvalue', 'mcols.qvalue'), digits = 3) 

feature_sum_annot %>% 
  dplyr::select(feature, external_gene_name, description, mcols.meth.diff, mcols.pvalue, mcols.qvalue, 
                seqnames, feature_strand, location, start, end) %>%
  write_xlsx(path = paste0("./tables/DMR_", file_path_name, ".xlsx"))

```
:::

## Graphs

### Volcano Plot

```{r}
#| echo: true
#| eval: true 
#| file: code/volcano_plot.R
#| warning: false
#| collapse: false
#| output: true
#| fig-width: 9
#| fig-height: 9
```

### Circular Plot

Shows the distribution of `r analysis_object` with q-value \< `r  qvalue_cut` along the chromosomes of X. tropicalis.

```{r}
#| echo: true
#| eval: true 
#| file: code/circular_plot.R
#| warning: false
#| collapse: false
#| output: true
#| message: false
#| fig-width: 8
#| fig-height: 8
#| fig-dpi: 300
```

### Genomic Regions

```{r}
#| echo: true
#| eval: true
#| fig-width: 7
#| fig-height: 7
#| warning: false
#| message: false

library(gt)

gen_regions %>% gt

plot_genomic_location()
```

# Gene Ontology analysis

```{r}
#| echo: true
#| eval: true
#| error: true
#| file: code/gene_ontology_analysis.R
#| warning: false
#| collapse: false
#| output: false
```

::: {.callout-note collapse="true"}
# GO tables

## GO Biological Function
**All**
```{r}
#| error: true

ggo_table_all
```

**Hypermethylated**
```{r}
#| error: true

ggo_table_hyper
```

**Hypomethylated**
```{r}
#| error: true

ggo_table_hypo
```
## GO Over-representation

**All**

```{r}
#| error: true

go_overrep_table_all
```

**Hypermethylated**
```{r}
#| error: true

go_overrep_table_hyper
```

**Hypomethylated**
```{r}
#| error: true

go_overrep_table_hypo
```


:::

**All**

```{r}
#| echo: true
#| eval: true
#| fig-width: 8
#| fig-height: 8
#| error: true

ggo_plot_all

go_overrep_barplot_all
```

**Hypermethylated**

```{r}
#| echo: true
#| eval: true
#| fig-width: 8
#| fig-height: 8
#| error: true

ggo_plot_hyper

go_overrep_barplot_hyper
```

**Hypomethylated**

```{r}
#| echo: true
#| eval: true
#| fig-width: 8
#| fig-height: 8
#| error: true

ggo_plot_hypo

go_overrep_barplot_hypo
```
# KEGG analysis

```{r}
#| echo: true
#| eval: true
#| error: true
#| file: code/kegg_analysis.R
#| warning: false
#| collapse: false
#| output: false
```

::: {.callout-note collapse="true"}
# KEGG table

## KEGG Over-representation **All**

```{r}
#| error: true

df_kegg_overrep_table_all
```

## **Hypermethylated**

```{r}
#| error: true

df_kegg_overrep_table_hyper
```

## **Hypomethylated**

```{r}
#| error: true

df_kegg_overrep_table_hypo
```

:::

## KEGG Over-representation **All**

```{r}
#| echo: true
#| eval: true
#| fig-width: 8
#| fig-height: 8
#| error: true


kegg_overrep_barplot_all
```

## KEGG Over-representation **Hypermethylated**

```{r}
#| echo: true
#| eval: true
#| fig-width: 8
#| fig-height: 8
#| error: true


kegg_overrep_barplot_hyper
```

## KEGG Over-representation **Hypomethylated**

```{r}
#| echo: true
#| eval: true
#| fig-width: 8
#| fig-height: 8
#| error: true


kegg_overrep_barplot_hypo
```